src/adagrams.js

```diff
@@ -1,15 +1,187 @@
+};
+// my copy
+
+// helper function to build weighted letter pool
+const makeWeightedPool = () => {
+  const weightedPool = [];
+  for (let letter in LETTER_POOL) {
+    let letterFreq = LETTER_POOL[letter];
+    // push letter to weightedPool, repeat letterFreq times
+    weightedPool.push(...letter.repeat(letterFreq));
+  }
+  return weightedPool;
+};
+
+////////// WAVE 1 //////////
 export const drawLetters = () => {
-  // Implement this method for wave 1
+  // call helper function
+  let weightedPool = makeWeightedPool();
+
+  // fill randomIndices with 10 random indices from weightedPool
+  const randomIndices = new Set();
+  while (randomIndices.size < 10) {
+    let randomIndex = Math.floor(Math.random() * weightedPool.length);
+    randomIndices.add(randomIndex);
+  }
+
+  // use the random indices to get letters
+  const lettersInHand = [];
+  for (let index of randomIndices) {
+    lettersInHand.push(weightedPool[index]);
+  }
+
+  return lettersInHand;
 };
 
+////////// WAVE 2 //////////
 export const usesAvailableLetters = (input, lettersInHand) => {
-  // Implement this method for wave 2
+  // change input from string to array
+  const lettersPlayed = input.split("");
+
+  // make a copy of our hand
+  const lettersInHandCopy = [...lettersInHand];
+
+  // if letter is in our hand, remove it. if it isn't, return false
+  for (let letter of lettersPlayed) {
+    if (lettersInHandCopy.includes(letter)) {
+      // remove one letter from our hand, then break out
+      for (let i = 0; i < lettersInHandCopy.length; i++) {
+        if (lettersInHandCopy[i] === letter) {
+          lettersInHandCopy.splice(i, 1);
+          break;
+        }
+      }
+    } else {
+      return false;
+    }
+  }
+
+  // if all letters were in our hand, return true
+  return true;
 };
 
+////////// WAVE 3 //////////
 export const scoreWord = (word) => {
-  // Implement this method for wave 3
+  // change input from string to array
+  const wordLetters = word.split("");
+
+  // get the score of each letter and add to wordScore
+  let wordScore = 0;
+  for (let letter of wordLetters) {
+    // get letter score from upper- or lowercase letters
+    let letterScore = LETTER_SCORES[letter.toUpperCase()];
+    wordScore = wordScore + letterScore;
+  }
+
+  // add length bonus to wordScore
+  if (wordLetters.length >= 7) {
+    wordScore = wordScore + 8;
+  }
+
+  return wordScore;
 };
 
+////////// WAVE 4 ///////////
 export const highestScoreFrom = (words) => {
-  // Implement this method for wave 4
+  // find highest score
+  let highScore = 0;
+  for (let word of words) {
+    // call function from wave 3
+    let wordScore = scoreWord(word);
+    if (wordScore > highScore) {
+      highScore = wordScore;
+    }
+  }
+
+  // make a list of all words w/ highest score
+  let highestScoringWords = [];
+  for (let word of words) {
+    let wordScore = scoreWord(word);
+    if (wordScore === highScore) {
+      highestScoringWords.push(word);
+    }
+  }
```

Comment:
This works but you could simplify this by creating one variable to hold the high scoring data. Using two variables takes up more memory and most primitives (booleans, numbers, null, undefined, etc.) are constant (O(1)) space. This is because, no matter if true or false, or 1 or 1,000, the amount of space consumed stays constant. Reference types (arrays and objects) — Reference types require O(n) space as well. ’n’ being the length for an array, or in the case of an object, ’n’ being the amount of keys. Reference types are O(n) because the more elements in an array, or the more keys in an object, the more space they consume. 

The `highestScoringWords` array does have a space complexity of 0(n), while the `highScore` number has a space complexity of 0(1), but since that's less than the `highestScoringWords` variable, we drop it. You are also using a `wordScore` variable which holds the result of scoreWord call and you are creating this variable twice, which triggers the `scoreWord` function call twice.